### PR TITLE
Refactor fix-rules.py tool

### PR DESCRIPTION
**Pull request by ggbecker**
#### Description:

- This PR aims to refactor fix-rules.py so it works with the new directory structure. Additionally it fixes some faulty JINJA macros when an error occurs.

#### Rationale:
This
Utility for fixing mistakes in .rule files
Commands:
	- empty_identifiers - check and fix rules with empty identifiers
	- prefixed_identifiers - check and fix rules with prefixed (CCE-) identifiers
	- invalid_identifiers - check and fix rules with invalid identifiers
	- int_identifiers - check and fix rules with pseudo-integer identifiers
	- empty_references - check and fix rules with empty references
	- int_references - check and fix rules with pseudo-integer references

### CODE REVIEW

File: utils/fix-rules.py [hunk 1]
```diff
@@ -91,32 +92,31 @@ def find_rules(directory, func):
...
+            # for d in dirs:
+            #     product_yamls[os.path.join(root, d)] = product_yaml
+            #     product_yaml_paths[os.path.join(root, d)] = product_yaml_path
         else:
             pass
 
         for filename in files:
             path = os.path.join(root, filename)
-            if len(path) < 5 or path[-5:] != '.rule':
+            if len(path) < 8 or path[-8:] != 'rule.yml' or "tests/" in path:
```
Comment (cipherboy) [inline utils/fix-rules.py]: One thought...

What about providing a content directory walker in the `ssg` module. Something like:

```
for item in ssg.walk_content(types=["product", "rule"]):
    pass
```

Or something similar. Probably out of scope for this PR, but I think it is a common enough occurrence that abstracting away the logic might be nice. Either leave it as a filesystem path or parse the YAML or something... My 2c.

Comment (jan-cerny) [reply utils/fix-rules.py]: Also, please remove magic constants like -8.

**Comment by jan-cerny** [comment]
Hi @ggbecker , I'm not able to get it working. In all commands I got a traceback. For example:
```
$ python3 utils/fix-rules.py empty_identifiers /home/jcerny/scap-security-guide
Exception while handling file: /home/jcerny/scap-security-guide/linux_os/guide/services/ntp/service_chronyd_or_ntpd_enabled/rule.yml
Traceback (most recent call last):
  File "utils/fix-rules.py", line 515, in <module>
    __main__()
  File "utils/fix-rules.py", line 500, in __main__
    fix_empty_identifiers(args.ssg_root)
  File "utils/fix-rules.py", line 387, in fix_empty_identifiers
    results = find_rules(directory, has_empty_identifier)
  File "utils/fix-rules.py", line 125, in find_rules
    if func(path, product_yaml):
  File "utils/fix-rules.py", line 13, in has_empty_identifier
    rule = yaml.open_and_macro_expand(yaml_file, product_yaml)
  File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 124, in open_and_macro_expand
    return open_and_expand(yaml_file, substitutions_dict)
  File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 102, in open_and_expand
    yaml_contents = _open_yaml(expanded_template, yaml_file, substitutions_dict)
  File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 69, in _open_yaml
    raise e
  File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 46, in _open_yaml
    yaml_contents = yaml.load(stream, Loader=yaml_SafeLoader)
  File "/usr/lib64/python3.7/site-packages/yaml/__init__.py", line 114, in load
    return loader.get_single_data()
  File "/usr/lib64/python3.7/site-packages/yaml/constructor.py", line 41, in get_single_data
    node = self.get_single_node()
  File "ext/_yaml.pyx", line 707, in _yaml.CParser.get_single_node
  File "ext/_yaml.pyx", line 725, in _yaml.CParser._compose_document
  File "ext/_yaml.pyx", line 776, in _yaml.CParser._compose_node
  File "ext/_yaml.pyx", line 892, in _yaml.CParser._compose_mapping_node
  File "ext/_yaml.pyx", line 905, in _yaml.CParser._parse_next_event
yaml.parser.ParserError: while parsing a block mapping
  in "<unicode string>", line 1, column 1
did not find expected key
  in "<unicode string>", line 10, column 5
```

**Comment by ggbecker** [comment]
> Hi @ggbecker , I'm not able to get it working. In all commands I got a traceback. For example:
> 
> ```
> $ python3 utils/fix-rules.py empty_identifiers /home/jcerny/scap-security-guide
> Exception while handling file: /home/jcerny/scap-security-guide/linux_os/guide/services/ntp/service_chronyd_or_ntpd_enabled/rule.yml
> Traceback (most recent call last):
>   File "utils/fix-rules.py", line 515, in <module>
>     __main__()
>   File "utils/fix-rules.py", line 500, in __main__
>     fix_empty_identifiers(args.ssg_root)
>   File "utils/fix-rules.py", line 387, in fix_empty_identifiers
>     results = find_rules(directory, has_empty_identifier)
>   File "utils/fix-rules.py", line 125, in find_rules
>     if func(path, product_yaml):
>   File "utils/fix-rules.py", line 13, in has_empty_identifier
>     rule = yaml.open_and_macro_expand(yaml_file, product_yaml)
>   File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 124, in open_and_macro_expand
>     return open_and_expand(yaml_file, substitutions_dict)
>   File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 102, in open_and_expand
>     yaml_contents = _open_yaml(expanded_template, yaml_file, substitutions_dict)
>   File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 69, in _open_yaml
>     raise e
>   File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 46, in _open_yaml
>     yaml_contents = yaml.load(stream, Loader=yaml_SafeLoader)
>   File "/usr/lib64/python3.7/site-packages/yaml/__init__.py", line 114, in load
>     return loader.get_single_data()
>   File "/usr/lib64/python3.7/site-packages/yaml/constructor.py", line 41, in get_single_data
>     node = self.get_single_node()
>   File "ext/_yaml.pyx", line 707, in _yaml.CParser.get_single_node
>   File "ext/_yaml.pyx", line 725, in _yaml.CParser._compose_document
>   File "ext/_yaml.pyx", line 776, in _yaml.CParser._compose_node
>   File "ext/_yaml.pyx", line 892, in _yaml.CParser._compose_mapping_node
>   File "ext/_yaml.pyx", line 905, in _yaml.CParser._parse_next_event
> yaml.parser.ParserError: while parsing a block mapping
>   in "<unicode string>", line 1, column 1
> did not find expected key
>   in "<unicode string>", line 10, column 5
> ```

Can't reproduce it. Can you run with python2 and post the result?

**Comment by ggbecker** [comment]
@jan-cerny I supect it might be related to the introduction of this str conversion: `str(i_value)):` from this commit 43cd7bd153b4901a91e7e5aa2cc776b47632f8e2

**Comment by jan-cerny** [comment]
@ggbecker 
```
$ python2 utils/fix-rules.py empty_identifiers /home/jcerny/scap-security-guide
Exception while handling file: /home/jcerny/scap-security-guide/linux_os/guide/services/ntp/service_chronyd_or_ntpd_enabled/rule.yml
Traceback (most recent call last):
  File "utils/fix-rules.py", line 515, in <module>
    __main__()
  File "utils/fix-rules.py", line 500, in __main__
    fix_empty_identifiers(args.ssg_root)
  File "utils/fix-rules.py", line 387, in fix_empty_identifiers
    results = find_rules(directory, has_empty_identifier)
  File "utils/fix-rules.py", line 125, in find_rules
    if func(path, product_yaml):
  File "utils/fix-rules.py", line 13, in has_empty_identifier
    rule = yaml.open_and_macro_expand(yaml_file, product_yaml)
  File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 124, in open_and_macro_expand
    return open_and_expand(yaml_file, substitutions_dict)
  File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 102, in open_and_expand
    yaml_contents = _open_yaml(expanded_template, yaml_file, substitutions_dict)
  File "/home/jcerny/scap-security-guide/ssg/yaml.py", line 69, in _open_yaml
    raise e
yaml.parser.ParserError: while parsing a block mapping
  in "<unicode string>", line 1, column 1
did not find expected key
  in "<unicode string>", line 10, column 5
```

**Comment by jan-cerny** [comment]
I don't experience my problem anymore. I have tried it with both Python 2 and 3.

**Pull request merged**